engine/execution/state/state.go

```diff
@@ -19,10 +18,65 @@ import (
 	"github.com/onflow/flow-go/storage/badger/procedure"
 )
 
-// ReadOnlyExecutionState allows to read the execution state
-type ReadOnlyExecutionState interface {
-	// NewStorageSnapshot creates a new ready-only view at the given state commitment.
-	NewStorageSnapshot(flow.StateCommitment) fvmState.StorageSnapshot
+// SnapshotReader provides read functionality at a specific snapshot
+//
+// TODO(ramtin): update this interface with ReadByPrefix and Warmup functionality
+// TODO(ramtin): maybe consolidate this with the FVM one
+type SnapshotReader interface {
+	// Read returns the value for a single register
+	// if no value is found it returns an empty byte slice and no error
+	// returned errors are fatal
+	// TODO(ramtin): rename to Read
+	Get(id flow.RegisterID) (value flow.RegisterValue, err error)
+
+	// BatchRead returns a set of values for the given registerIDs
+	// it has the same behaviour as `Read` except it operates on a batch of registers
+	// returned errors are fatal
+	BatchRead(ids []flow.RegisterID) (itr flow.RegisterIterator, err error)
+}
+
+type BlockAwareStorage interface {
+	// OnFinalizedBlock is called whenever a block has been finalized.
+	// Calls are in the order the blocks are finalized.
+	// Prerequisites:
+	// Implementation must be concurrency safe; Non-blocking;
+	// and must handle repetition of the same events (with some processing overhead).
+	OnFinalizedBlock(block *flow.Header)
+
+	// OnSealedBlock is called when a block is sealed
+	// Calls are in the order blocks get sealed.
+	// Prerequisites:
+	// Implementation must be concurrency safe; Non-blocking;
+	// and must handle repetition of the same events (with some processing overhead).
+	OnSealedBlock(block *flow.Header)
+
+	// OnBlockExecuted is called when a block is executed
+	// Calls are in the order blocks get executed.
+	// Prerequisites:
+	// Implementation must be concurrency safe;
+	// and must handle repetition of the same events (with some processing overhead).
+	OnBlockExecuted(block *flow.Header, result *execution.ComputationResult) error
+}
+
+type ReadOnlyRegisterStorage interface {
+	module.ReadyDoneAware
+	// SnapshotReaderAtBlock returns an SnapshotReader for the given blockID
+	// if a register is not changed in the given block the latest value from ancestor blocks is returned
+	// it could be used for reading several values of the same blockID
+	SnapshotReaderAtBlock(blockID flow.Identifier) (SnapshotReader, error)
+
+	// TODO(ramtin): consolidate this into above and depricate
+	NewStorageSnapshot(commitment flow.StateCommitment) SnapshotReader
+}
+
+// RegisterStorage stores registers at given height,
```

Comment:
| at given height

Should be "at given block", right? we need to be fork-aware